src/inmanta/env.py

```diff
@@ -182,26 +237,207 @@ def _run_command_and_log_output(
+                    version = item.specs
+                    marker = item.marker
+                    if hasattr(item, "url"):
+                        url = item.url
+            except InvalidRequirement:
+                url = req_spec
+
+            if name not in modules:
+                modules[name] = {"name": name, "version": [], "markers": []}
+
+            if version is not None:
+                modules[name]["version"].extend(version)
+
+            if marker is not None:
+                modules[name]["markers"].append(marker)
+
+            if url is not None:
+                modules[name]["url"] = url
+
+        requirements_file = ""
+        for module, info in modules.items():
+            version_spec = ""
+            markers: str = ""
+            if len(info["version"]) > 0:
+                version_spec = " " + (", ".join(["%s %s" % (a, b) for a, b in info["version"]]))
+
+            if len(info["markers"]) > 0:
+                markers = " ; " + (" and ".join(map(str, info["markers"])))
+
+            if "url" in info:
+                module = info["url"]
+
+            requirements_file += module + version_spec + markers + "\n"
+
+        return requirements_file
+
+    def _install(self, requirements_list: List[str]) -> None:
+        """
+        Install requirements in the given requirements file
+        """
+        content_requirements_file = self._gen_content_requirements_file(requirements_list)
+        with requirements_txt_file(content=content_requirements_file) as requirements_file:
+            with requirements_txt_file(content=self._get_constraint_on_inmanta_package()) as constraint_file:
+                cmd: List[str] = PipCommandBuilder.compose_install_command(
+                    python_path=self.python_path,
+                    requirements_files=[requirements_file],
+                    constraints_files=[constraint_file],
+                )
+                try:
+                    self._run_command_and_log_output(cmd, stderr=subprocess.STDOUT)
+                except Exception:
+                    LOGGER.error("requirements: %s", content_requirements_file)
+                    raise
+
+    def _read_current_requirements_hash(self) -> str:
+        """
+        Return the hash of the requirements file used to install the current environment
+        """
+        path = os.path.join(self.env_path, "requirements.sha1sum")
+        if not os.path.exists(path):
+            return ""
+
+        with open(path, "r", encoding="utf-8") as fd:
+            return fd.read().strip()
+
+    def _set_current_requirements_hash(self, new_hash: str) -> None:
+        """
+        Set the current requirements hash
+        """
+        path = os.path.join(self.env_path, "requirements.sha1sum")
+        with open(path, "w+", encoding="utf-8") as fd:
+            fd.write(new_hash)
+
+    def install_from_list(self, requirements_list: List[str], detailed_cache: bool = False, cache: bool = True) -> None:
```

Comment:
I think @sanderr remarked that this can no longer be relied on, as the venv is now shared, and that it is beter to check the constraints against the master set (this check may already be implemented)

Reply:
More specifically, the cache file can not be relied upon because the venv is not fully under our control. `ActiveEnv.check` does some checking against the master set. It doesn't completely match with what we need here but it might be useful as a reference.